docs/rfcs/0001-image-watcher.md

```diff
@@ -0,0 +1,82 @@
+- Start Date: 2020-11-18
+- Target Version: 1.0
+
+# Summary
+This RFC proposes adding a new feature to trigger Deployment when an image tag stored at a container registry is updated.
+Similar to Flux's [Image Scan](https://docs.fluxcd.io/en/1.19.0/faq/#how-often-does-flux-check-for-new-images).
+
+# Motivation
+The canonical deployment flow with PipeCD is:
+
+1. CI server pushes the generated image to the container registry after app-repo updated.
+2. User updates the config-repo manually.
+
+It is the User's responsibility to automate these 1 and 2 step to be done in a series of actions, while it is a bit quite painful.
+Therefore, we would like to solve this problem by letting Piped monitor the container registry.
+
+# Detailed design
+Piped compares the image tags defined in git with the latest tags stored in the container registry, and then pushes them to git if there are any deviations.
+
+### Usage
+Note: Where, we call the client that accesses the container registry to `Image Provider`.
+
+First of all, we need to register the credentials for accessing the image provider with Piped.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: Piped
+spec:
+  imageProviders:
+    - name: my-dockerhub
+      type: DOCKERHUB
+      # Indicates how many minutes Piped polls every.
+      interval: 5m
+      config:
+        username: foo
+        passwordFile: /etc/piped-secret/dockerhub-pass
+```
+
+We define the information about image tag underneath `.pipe/` directory:
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: ImageWatcher
+spec:
+  targets:
+    - image: gcr.io/pipecd/foo
+      provider: my-gcr
+      filePath: foo/deployment.yaml
+      field: spec.containers[0].image
+    - image: pipecd/bar
+      provider: my-dockerhub
+      filePath: bar/deployment.yaml
+      field: spec.containers[0].image
+```
+
+### Architecture
+`Image Provider` pulls respectively periodically.
+`Image Watcher` pulls from the git repo, compares respectively and pushes the one with differences.
+
+![](assets/image-watcher.jpg)
+
+
+# Alternatives
+Initialy, I was thinking it should be defined in Deployment Configuration, but it's slightly different from configuration about Deployment.
+In addition to that, it is too complicated to inspect the entire git path every time.
+
+```yaml
+apiVersion: pipecd.dev/v1beta1
+kind: KubernetesApp
+spec:
+  imageWatcher:
+    targets:
+    - image: gcr.io/pipecd/helloworld
+      path:
+        filename: deployment.yaml
+        field: spec.containers[0].image
+      provider: my-dockerhub
+```
+
+# Unresolved questions
+
+- How to update template. Currently considering using https://github.com/vmware-tanzu/carvel-ytt
```

Comment:
As far as I know, the main target of `carvel-ytt` is providing a tool for end-user, not a library for developers. So importing its go-package maybe not a good choice. They can make any change to those packages in the future.

Reply:
I was attempting to execute it as a child process of piped, but anyway still not enough investigation at all.

Reply:
Rather than keeping unresolved questions as is, should we merge after investigating enough?

Reply:
>I was attempting to execute it as a child process of piped, but anyway still not enough investigation at all.

If possible, we should avoid running those tools by using `cmd`. Because it requires installing the tool beside our piped binary. The user should install only a single `piped` to use.

Reply:
>Rather than keeping unresolved questions as is, should we merge this RFC after investigating enough?

I am not sure but I think the `Unresolved Questions` section is for storing the questions that still not be resolved after our *discussion* here.

Reply:
I got it. I will glad to investigate more 👍